app/server/lib/OIDCConfig.ts

```diff
@@ -0,0 +1,192 @@
+ * When running on localhost and http, the settings tested were with:
+ *   - GRIST_OIDC_SP_HOST=http://localhost:8484 (or whatever port you use for Grist)
+ *   - GRIST_OIDC_IDP_ISSUER=http://localhost:8080/realms/myrealm (replace 8080 by the port you use for keycloak)
+ *   - GRIST_OIDC_IDP_CLIENT_ID=myclient
+ *   - GRIST_OIDC_IDP_CLIENT_SECRET=YOUR_SECRET (as set in keycloak)
+ *   - GRIST_OIDC_IDP_SCOPES="openid email profile"
+ */
+
+import * as express from 'express';
+import {GristLoginSystem, GristServer} from './GristServer';
+import { Client, generators, Issuer, UserinfoResponse } from 'openid-client';
+import {Sessions} from './Sessions';
+import log from 'app/server/lib/log';
+import {Permit} from './Permit';
+
+const CALLBACK_URL = '/oauth2/callback';
+const LOGIN_URL = '/oauth2/login';
+
+const LOGIN_ACTION = 'oidc-login';
+const WAIT_IN_MINUTES = 20;
+
+export class OIDCConfig {
+  private _client: Client;
+
+  public constructor(private _gristServer: GristServer) {
+  }
+
+  public async initOIDC(): Promise<void> {
+    if (!process.env.GRIST_OIDC_SP_HOST) { throw new Error('initOIDC requires GRIST_OIDC_SP_HOST to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_ISSUER) { throw new Error('initOIDC requires GRIST_OIDC_IDP_ISSUER to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_ID) {
+      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_ID to be set');
+    }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_SECRET) {
+      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_SECRET to be set');
+    }
+    const spHost: string = process.env.GRIST_OIDC_SP_HOST;
+    const issuer = await Issuer.discover(process.env.GRIST_OIDC_IDP_ISSUER);
+    this._client = new issuer.Client({
+      client_id: process.env.GRIST_OIDC_IDP_CLIENT_ID,
+      client_secret: process.env.GRIST_OIDC_IDP_CLIENT_SECRET,
+      redirect_uris: [ new URL(CALLBACK_URL, spHost).href ],
+      response_types: ['code'],
+    });
+  }
+
+  public addEndpoints(app: express.Application, sessions: Sessions): void {
+    app.get(LOGIN_URL, this.handleLogin.bind(this));
+    app.get(CALLBACK_URL, this.handleCallback.bind(this, sessions));
+  }
+
+  public async handleCallback(sessions: Sessions, req: express.Request, res: express.Response): Promise<void> {
+
+    try {
+      const params = this._client.callbackParams(req);
+      const { state } = params;
+      if (!state) {
+        throw new Error('Login or logout failed to complete');
+      }
+      const codeVerifier = await this._retrieveCodeVerifierFromPermit(state);
+      const tokenSet = await this._client.callback(
+        new URL(CALLBACK_URL, process.env.GRIST_OIDC_SP_HOST).href,
+        params,
+        { state, code_verifier: codeVerifier }
+      );
+      const userInfo = await this._client.userinfo(tokenSet);
+      const profile = this._makeUserProfileFromUserInfo(userInfo);
+      const scopedSession = sessions.getOrCreateSessionFromRequest(req);
```

Comment:
In the SAML implementation, there is an extra argument `{sessionId: state.sessionId}`, with the explanation that the cookie may not be included in the callback request (when it's redirected from the issuer). Maybe this isn't needed here because the callback is a GET request (rather than POST), but just wanted to confirm.

Reply:
Right. In my case, extracting the sessionId worked well using the cookie, but after your message, I am a bit unsure.

I was about to add a check to ensure that the sessionId matches between the permit and the cookie.

However, if we cannot rely on the cookie, if I am correct, that would mean that the code challenge can be retrieved only thanks to the value of the `state` variable, which means that the code challenge is not much useful here (protection offered by the PKCE is explained here: <https://dzone.com/articles/what-is-pkce>)

I am looking for a way to store the code verifier in the session, but it appears that at the contrary of some other session mechanisms, I have to retrieve some randomly generated key and store it by myself. That's the reason why I am passing this random key as the state, but I am not very confident of that way of doing thinks. If you have a better idea, I would be grateful :pray:

Reply:
PKCE does seem recommended for security, so probably best to keep that.

Looking again at the issue with the cookie being available, it seems that the reason we had to switch to using a `state` variable with SAML was that the redirect could be a POST, and SameSite=Lax cookies would not be included. With OIDC, the redirect is a GET, and such cookies ARE included. So actually, we should have access to the session.

Basically, this means we should be able to store the `code_verifier` in the session, and may not need the Permit method at all. The benefit would be to remove the somewhat arbitrary 20-minute timeout, and a bit of simplification. It would be good for @paulfitz to check -- maybe there are reasons to prefer the Permit-based approach.

If something isn't working with sessions, it may be for an unrelated reason. I forget the details, but sometimes we've needed to include a call to `forceSessionChange()` to get session updates to get saved.

Reply:
> I hadn't (and haven't) looked at exactly what is being stored here, and whether it needs to be kept opaque across some round trip through another service

The `codeVerifier` is some random values that is SHA256-hashed and passed during the login phase as the `codeChallenge`, and that should be stored so it can be passed in plain text during the callback phase. The authentication server then compare whether `codeChallenge === sha256(codeVerifier)`.

I hope that's a clear explanation. If not, I am a bit in a hurry, but I can provide provide more information tomorrow.